programs/whirlpool/src/util/v2/token.rs

```diff
@@ -0,0 +1,378 @@
+use crate::state::{token_badge, Whirlpool};
+use crate::errors::ErrorCode;
+use anchor_lang::prelude::*;
+use anchor_spl::token_2022::spl_token_2022::extension::transfer_fee::{TransferFee, MAX_FEE_BASIS_POINTS};
+use anchor_spl::token_interface::spl_token_2022::extension::BaseStateWithExtensions;
+
+use anchor_spl::token::Token;
+use anchor_spl::token_2022::spl_token_2022::{self, extension::{self, StateWithExtensions}, state::AccountState};
+use anchor_spl::token_interface::{Mint, TokenAccount, TokenInterface};
+use anchor_spl::memo::{self, Memo, BuildMemo};
+use spl_transfer_hook_interface;
+
+use token_badge::TokenBadge;
+
+pub fn transfer_from_owner_to_vault_v2<'info>(
+    authority: &Signer<'info>,
+    token_mint: &InterfaceAccount<'info, Mint>,
+    token_owner_account: &InterfaceAccount<'info, TokenAccount>,
+    token_vault: &InterfaceAccount<'info, TokenAccount>,
+    token_program: &Interface<'info, TokenInterface>,
+    memo_program: &Program<'info, Memo>,
+    transfer_hook_accounts: &Option<Vec<AccountInfo<'info>>>,
+    amount: u64,
+) -> Result<()> {
+    // TransferFee extension
+    if let Some(epoch_transfer_fee) = get_epoch_transfer_fee(token_mint)? {
+        // log applied transfer fee
+        // - Not must, but important for ease of investigation and replay when problems occur
+        // - Use Memo because logs risk being truncated
```

Comment:
This requires extra compute units and logs to the transaction log anyway (which is the same as just doing `msg!` ourselves? In Solana v1.18 transactions that require less CUs get more prio so there is now an incentive to keep the CUs as low as reasonably possible

Reply:
Logs may be truncated, but instructions that remain as innerInstruction are never truncated and we can "data" as log.
This is the advantage of using Memo.

I looked into this issue again and found that the following noop program is used in account-compression.
It appears to consume much less CU than Memo, but the down side is to make sure that the noop_program is received in the v2 instruction. (extra 32 bytes without ALT)

https://solana.stackexchange.com/questions/10090/cost-of-logging-vs-storing-data-in-solana-programs/10100#10100
https://crates.io/crates/spl-noop
no-op program: https://solscan.io/account/noopb9bkMVfRPU8AsbpTUg8AQkHtKwMYZiFUjNRtMmV

WDYT about the reason to use Memo and noop program ?

Reply:
Since Memo already needs to be there for the memo-required extension we might as well just use that. I think the impact on CUs will not be that large whereas having to add the noop_program everytime adds to the tx size which is currently sometimes already an issue (harvest, withdraw with ata creation, etc.)

Reply:
Thanks for sharing your thoughts.
Agreed. Let's move forward with Memo program.👍